### PR TITLE
Add ODR handling strategy to ASTImporter

### PR DESCRIPTION
**Pull request by balazske**

### CODE REVIEW

File: include/clang/AST/ASTImporter.h [hunk 1]
```diff
@@ -89,6 +89,8 @@ class TypeSourceInfo;
     using ImportedCXXBaseSpecifierMap =
         llvm::DenseMap<const CXXBaseSpecifier *, CXXBaseSpecifier *>;
 
+    enum class ODRHandlingType { Conservative, Liberal };
```
Comment (martong) [inline include/clang/AST/ASTImporter.h:92]: In LLVM we typically name these enums as ...Kind. So it should be ODRHandlingKind.

File: unittests/AST/ASTImporterFixtures.h [hunk 2]
```diff
@@ -174,6 +193,46 @@ class ASTImporterOptionSpecificTestBase
   ArgVector getExtraArgs() const override { return GetParam(); }
 };
 
+template <class T>
+::testing::AssertionResult isSuccess(llvm::Expected<T> &ValOrErr) {
+  if (ValOrErr)
+    return ::testing::AssertionSuccess() << "Expected<> contains not an error.";
```
Comment (martong) [inline unittests/AST/ASTImporterFixtures.h:199]: "... contains no error." ?

File: unittests/AST/ASTImporterGenericRedeclTest.cpp [hunk 3]
```diff
@@ -596,21 +665,36 @@ ASTIMPORTER_INSTANTIATE_TYPED_TEST_CASE(RedeclChain, VariableTemplate, ,
 ASTIMPORTER_INSTANTIATE_TYPED_TEST_CASE(RedeclChain, FunctionTemplateSpec, ,
                                         ImportPrototypeThenProtoAndDefinition)
 
-INSTANTIATE_TEST_CASE_P(ParameterizedTests, RedeclChainFunction,
+ASTIMPORTER_ODR_INSTANTIATE_TYPED_TEST_CASE(RedeclChain, Class, Conservative, ,
+                                            DoNotImportConflictingDefinition)
+ASTIMPORTER_ODR_INSTANTIATE_TYPED_TEST_CASE(RedeclChain, Class, Liberal, ,
```
Comment (martong) [inline unittests/AST/ASTImporterGenericRedeclTest.cpp:670]: Will we have the Liberal test cases instantiated for other kind of Decls too?

File: unittests/AST/ASTImporterFixtures.h [hunk 4]
```diff
@@ -174,6 +193,46 @@ class ASTImporterOptionSpecificTestBase
...
+  else
+    return ::testing::AssertionFailure()
+           << "Expected<> contains error: " << toString(ValOrErr.takeError());
+}
+
+template <class T>
+::testing::AssertionResult isImportError(llvm::Expected<T> &ValOrErr,
```
Comment (martong) [inline unittests/AST/ASTImporterFixtures.h:206]: :+1:

File: unittests/AST/ASTImporterFixtures.h [hunk 5]
```diff
@@ -174,6 +193,46 @@ class ASTImporterOptionSpecificTestBase
   ArgVector getExtraArgs() const override { return GetParam(); }
 };
 
+template <class T>
+::testing::AssertionResult isSuccess(llvm::Expected<T> &ValOrErr) {
```
Comment (martong) [inline unittests/AST/ASTImporterFixtures.h:197]: :+1:

File: include/clang/AST/ASTImporter.h (hunk 1 above)
Comment (balazske) [reply include/clang/AST/ASTImporter.h:92]: The description in http://llvm.org/docs/CodingStandards.html#name-types-functions-variables-and-enumerators-properly  is not clear, in this case it is not a subclass or union type so I was not sure.

File: unittests/AST/ASTImporterGenericRedeclTest.cpp (hunk 3 above)
Comment (balazske) [reply unittests/AST/ASTImporterGenericRedeclTest.cpp:670]: Yes, more tests follow. But there may be different cases for different types.

File: include/clang/AST/ASTImporter.h (hunk 1 above)
Comment (martong) [reply include/clang/AST/ASTImporter.h:92]: Ok.

**Review by martong — APPROVED** [review]

**Comment by balazske** [comment]
Can we revert this? I wanted to add new tests and there are some problems with the tests (in the current form it can be merged but not as it was previously). A new commit was after merge and does not appear now.

**Comment by martong** [comment]
Could you please create a new patch with the new tests and with the possible new changes?

**Pull request merged**